components/team_controller/team_controller.tsx

```diff
@@ -0,0 +1,248 @@
+    }, []);
+
+    // Effect runs on mount, add event listeners on windows object
+    useEffect(() => {
+        window.addEventListener('focus', handleFocus);
+        window.addEventListener('blur', handleBlur);
+        window.addEventListener('keydown', handleKeydown);
+
+        return () => {
+            window.removeEventListener('focus', handleFocus);
+            window.removeEventListener('blur', handleBlur);
+            window.removeEventListener('keydown', handleKeydown);
+        };
+    }, []);
+
+    // Effect runs on mount, adds active state to window
+    useEffect(() => {
+        const browserIsIosSafari = isIosSafari();
+        if (browserIsIosSafari) {
+            // Use iNoBounce to prevent scrolling past the boundaries of the page
+            iNoBounce.enable();
+        }
+
+        // Set up tracking for whether the window is active
+        window.isActive = true;
+
+        LocalStorageStore.setTeamIdJoinedOnLoad(null);
+
+        return () => {
+            window.isActive = false;
+
+            if (browserIsIosSafari) {
+                iNoBounce.disable();
+            }
+        };
+    }, []);
+
+    async function initTeamOrRedirect(team: Team) {
+        try {
+            await props.initializeTeam(team);
+            setTeam(team);
+        } catch (error) {
+            history.push('/error?type=team_not_found');
+        }
+    }
+
+    async function joinTeamOrRedirect(teamNameParam: string, joinedOnFirstLoad: boolean) {
+        try {
+            // skip reserved team names
+            if (Constants.RESERVED_TEAM_NAMES.includes(teamNameParam)) {
+                throw new Error('Team name is reserved');
+            }
+
+            const {data: joinedTeam} = await props.joinTeam(teamNameParam, joinedOnFirstLoad) as ActionResult<Team, ServerError>; // Fix in MM-46907;
+            if (joinedTeam) {
+                setTeam(joinedTeam);
+            } else {
+                throw new Error('Unable to join team');
+            }
+        } catch (error) {
+            history.push('/error?type=team_not_found');
+        }
+    }
+
+    const teamsListDependency = props.teamsList.map((team) => team.id).sort().join('+');
+
+    // Effect to run when url for team or teamsList changes
+    useEffect(() => {
+        // Prevents the RHS from closing when clicking on a global permalink.
+        emitCloseRightHandSide();
+
+        const teamFromURLParam = getTeamFromTeamList(props.teamsList, teamNameParam);
+        if (teamFromURLParam) {
+            // Team switch
+            initTeamOrRedirect(teamFromURLParam);
+            return;
+        }
+
+        if (teamNameParam && !team) {
+            // Team join on first load
+            joinTeamOrRedirect(teamNameParam, true);
+            return;
+        }
+
+        if (teamNameParam && team && team.name !== teamNameParam) {
+            // Team join by going through a link
+            joinTeamOrRedirect(teamNameParam, false);
+        }
```

Comment:
📝 Moved the business logic to actions makes it easier to test, and is inline with our MVC pattern